### PR TITLE
Add kclient functions for creating volumes

### PR DESCRIPTION
**Pull request by maysunfaisal**
**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:
Adds the base kclient package functions required for creating PVC and adds the Volume and Volume Mounts to the Pod Template Spec of the Devfile fat pod.

The logic to read the volumes from the Devfile components are in my personal branch and will open a PR for them once the PR for Devfile parser, as well as the PR for creating Deployment for the Devfile components are merged. 

Please have a look at my action plan as per issue comment https://github.com/openshift/odo/issues/2472#issuecomment-579966277. I have linked my personal POC branch changes for the Devfile volume logic against the checklist

**Which issue(s) this PR fixes**:

Partially implements https://github.com/openshift/odo/issues/2472

**How to test changes / Special notes to the reviewer**:
1. `cd pkg/kclient`
2. `go test -v` or `go test ./...`

```
maysuns-mbp:kclient maysun$ go clean -testcache && go test -v
=== RUN   TestCreateDeployment
=== RUN   TestCreateDeployment/Case:_Valid_deployment_name
=== RUN   TestCreateDeployment/Case:_Invalid_deployment_name
--- PASS: TestCreateDeployment (0.00s)
    --- PASS: TestCreateDeployment/Case:_Valid_deployment_name (0.00s)
    --- PASS: TestCreateDeployment/Case:_Invalid_deployment_name (0.00s)
=== RUN   TestGenerateContainer
=== RUN   TestGenerateContainer/#00
=== RUN   TestGenerateContainer/container1
--- PASS: TestGenerateContainer (0.00s)
    --- PASS: TestGenerateContainer/#00 (0.00s)
    --- PASS: TestGenerateContainer/container1 (0.00s)
=== RUN   TestGeneratePodTemplateSpec
=== RUN   TestGeneratePodTemplateSpec/podSpecTest
--- PASS: TestGeneratePodTemplateSpec (0.00s)
    --- PASS: TestGeneratePodTemplateSpec/podSpecTest (0.00s)
=== RUN   TestGeneratePVCSpec
=== RUN   TestGeneratePVCSpec/1
=== RUN   TestGeneratePVCSpec/2
--- PASS: TestGeneratePVCSpec (0.00s)
    --- PASS: TestGeneratePVCSpec/1 (0.00s)
    --- PASS: TestGeneratePVCSpec/2 (0.00s)
=== RUN   TestWaitAndGetPod
=== RUN   TestWaitAndGetPod/phase:_running
 ✓  Waiting for component to start [152677ns]
=== RUN   TestWaitAndGetPod/phase:_failed
 ✗  Waiting for component to start [92125ns]
=== RUN   TestWaitAndGetPod/phase:_unknown
 ✗  Waiting for component to start [62406ns]
--- PASS: TestWaitAndGetPod (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_running (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_failed (0.00s)
    --- PASS: TestWaitAndGetPod/phase:_unknown (0.00s)
=== RUN   TestCreatePVC
=== RUN   TestCreatePVC/Case:_Valid_pvc_name
=== RUN   TestCreatePVC/Case:_Invalid_pvc_name
--- PASS: TestCreatePVC (0.00s)
    --- PASS: TestCreatePVC/Case:_Valid_pvc_name (0.00s)
    --- PASS: TestCreatePVC/Case:_Invalid_pvc_name (0.00s)
=== RUN   TestAddPVCToPodTemplateSpec
=== RUN   TestAddPVCToPodTemplateSpec/podSpecTest
--- PASS: TestAddPVCToPodTemplateSpec (0.00s)
    --- PASS: TestAddPVCToPodTemplateSpec/podSpecTest (0.00s)
=== RUN   TestAddVolumeMountToPodTemplateSpec
=== RUN   TestAddVolumeMountToPodTemplateSpec/podSpecTest
=== RUN   TestAddVolumeMountToPodTemplateSpec/podSpecTest#01
--- PASS: TestAddVolumeMountToPodTemplateSpec (0.00s)
    --- PASS: TestAddVolumeMountToPodTemplateSpec/podSpecTest (0.00s)
    --- PASS: TestAddVolumeMountToPodTemplateSpec/podSpecTest#01 (0.00s)
=== RUN   TestGetPVCsFromSelector
=== RUN   TestGetPVCsFromSelector/Case:_Valid_pvc_name
--- PASS: TestGetPVCsFromSelector (0.00s)
    --- PASS: TestGetPVCsFromSelector/Case:_Valid_pvc_name (0.00s)
PASS
ok  	github.com/openshift/odo/pkg/kclient	0.371s
```

```
maysuns-mbp:kclient maysun$ go test ./...
ok  	github.com/openshift/odo/pkg/kclient	0.397s
```

### CODE REVIEW

File: pkg/kclient/generators.go [hunk 1]
```diff
@@ -58,3 +56,20 @@ func GenerateDeploymentSpec(podTemplateSpec corev1.PodTemplateSpec) *appsv1.Depl
...
+func GeneratePVCSpec(quantity resource.Quantity) *corev1.PersistentVolumeClaimSpec {
+
+	pvcSpec := &corev1.PersistentVolumeClaimSpec{
+		Resources: corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				corev1.ResourceStorage: quantity,
+			},
+		},
+		AccessModes: []corev1.PersistentVolumeAccessMode{
+			corev1.ReadWriteMany,
```
Comment (johnmcollier) [inline pkg/kclient/generators.go]: We can't use ReadWriteMany volumes due to the lack of Kube storage drivers that support it.

File: pkg/kclient/kclient.go [hunk 2]
```diff
@@ -49,3 +51,28 @@ func New() (*Client, error) {
...
+		Namespace:   namespace,
+		Labels:      labels,
+		Annotations: annotations,
+	}
+
+	return objectMeta
+}
+
+// GetPVCsFromSelector returns the PVCs based on the given selector
+func (c *Client) GetPVCsFromSelector(selector string) ([]corev1.PersistentVolumeClaim, error) {
```
Comment (johnmcollier) [inline pkg/kclient/kclient.go]: Can this go in the volumes.go file? It'll make modularizing the kclient package easier in the future.

File: pkg/kclient/kclient_test.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ktesting "k8s.io/client-go/testing"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestGetPVCsFromSelector(t *testing.T) {
```
Comment (johnmcollier) [inline pkg/kclient/kclient_test.go]: Can these be moved under the volumes_test.go?

File: pkg/kclient/generators.go (hunk 1 above)
Comment (maysunfaisal) [reply pkg/kclient/generators.go]: Updated to `RWO`

File: pkg/kclient/kclient.go (hunk 2 above)
Comment (maysunfaisal) [reply pkg/kclient/kclient.go]: Updated

File: pkg/kclient/kclient_test.go (hunk 3 above)
Comment (maysunfaisal) [reply pkg/kclient/kclient_test.go]: Updated

**Comment by girishramnani** [comment]
I am wrapped up in some other PRs so give me some time, I will get back to this by the end of this week.

**Comment by maysunfaisal** [comment]
Not sure if the Unit Test on MacOS X is flaky, because when I try locally on my Mac, they always seem to pass:

```
maysuns-mbp:odo maysun$ make test
go test -race github.com/openshift/odo/cmd/cli-doc github.com/openshift/odo/cmd/odo github.com/openshift/odo/pkg/application github.com/openshift/odo/pkg/application/labels github.com/openshift/odo/pkg/auth github.com/openshift/odo/pkg/catalog github.com/openshift/odo/pkg/component github.com/openshift/odo/pkg/component/labels github.com/openshift/odo/pkg/config github.com/openshift/odo/pkg/debug github.com/openshift/odo/pkg/devfile github.com/openshift/odo/pkg/devfile/parser github.com/openshift/odo/pkg/devfile/versions github.com/openshift/odo/pkg/devfile/versions/1.0.0 github.com/openshift/odo/pkg/devfile/versions/common github.com/openshift/odo/pkg/kclient github.com/openshift/odo/pkg/log github.com/openshift/odo/pkg/log/fidget github.com/openshift/odo/pkg/machineoutput github.com/openshift/odo/pkg/notify github.com/openshift/odo/pkg/occlient github.com/openshift/odo/pkg/odo/cli github.com/openshift/odo/pkg/odo/cli/application github.com/openshift/odo/pkg/odo/cli/catalog github.com/openshift/odo/pkg/odo/cli/catalog/describe github.com/openshift/odo/pkg/odo/cli/catalog/list github.com/openshift/odo/pkg/odo/cli/catalog/search github.com/openshift/odo/pkg/odo/cli/catalog/util github.com/openshift/odo/pkg/odo/cli/component github.com/openshift/odo/pkg/odo/cli/component/ui github.com/openshift/odo/pkg/odo/cli/config github.com/openshift/odo/pkg/odo/cli/debug github.com/openshift/odo/pkg/odo/cli/login github.com/openshift/odo/pkg/odo/cli/logout github.com/openshift/odo/pkg/odo/cli/preference github.com/openshift/odo/pkg/odo/cli/project github.com/openshift/odo/pkg/odo/cli/service github.com/openshift/odo/pkg/odo/cli/service/ui github.com/openshift/odo/pkg/odo/cli/storage github.com/openshift/odo/pkg/odo/cli/ui github.com/openshift/odo/pkg/odo/cli/url github.com/openshift/odo/pkg/odo/cli/utils github.com/openshift/odo/pkg/odo/cli/version github.com/openshift/odo/pkg/odo/genericclioptions github.com/openshift/odo/pkg/odo/util github.com/openshift/odo/pkg/odo/util/completion github.com/openshift/odo/pkg/odo/util/validation github.com/openshift/odo/pkg/preference github.com/openshift/odo/pkg/project github.com/openshift/odo/pkg/secret github.com/openshift/odo/pkg/service github.com/openshift/odo/pkg/storage github.com/openshift/odo/pkg/storage/labels github.com/openshift/odo/pkg/testingutil github.com/openshift/odo/pkg/testingutil/filesystem github.com/openshift/odo/pkg/url github.com/openshift/odo/pkg/url/labels github.com/openshift/odo/pkg/util github.com/openshift/odo/pkg/version
?   	github.com/openshift/odo/cmd/cli-doc	[no test files]
?   	github.com/openshift/odo/cmd/odo	[no test files]
ok  	github.com/openshift/odo/pkg/application	(cached)

..

ok  	github.com/openshift/odo/pkg/url/labels	(cached)
ok  	github.com/openshift/odo/pkg/util	(cached)
?   	github.com/openshift/odo/pkg/version	[no test files]
maysuns-mbp:odo maysun$ 
```

File: pkg/kclient/volumes_test.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.podName, func(t *testing.T) {
+
+			objectMeta := CreateObjectMeta(tt.podName, tt.namespace, tt.labels, nil)
+
+			podTemplateSpec := GeneratePodTemplateSpec(objectMeta, tt.serviceAccount, []corev1.Container{*container})
+
+			_ = AddVolumeMountToPodTemplateSpec(podTemplateSpec, tt.volumeName, tt.pvc, tt.containerMountPathsMap)
```
Comment (girishramnani) [inline pkg/kclient/volumes_test.go]: do we need the `_ =` assignment here?

**Comment by kadel** [comment]
> Not sure if the Unit Test on MacOS X is flaky, because when I try locally on my Mac, they always seem to pass:

In is flaky test :-( https://github.com/openshift/odo/issues/2588

File: pkg/kclient/volumes_test.go (hunk 4 above)
Comment (kadel) [reply pkg/kclient/volumes_test.go]: The function returns an error, so it should probably be checked here.

Comment (maysunfaisal) [reply pkg/kclient/volumes_test.go]: I have updated the test, I am now testing the err condition as well. I missed it previously. I have also updated some test code to reuse a odo util function for labels.

Pls review at your convenience, thx!

File: pkg/kclient/volumes_test.go [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+			PVCs, err := fkclient.GetPVCsFromSelector(selector)
+			if !tt.wantErr && err != nil {
+				t.Errorf("TestGetPVCsFromSelector: Error listing PVCs with selector: %v", err)
+			}
+
+			if len(PVCs) == 0 || len(PVCs) > 1 {
+				if !tt.wantErr {
+					t.Errorf("TestGetPVCsFromSelector: Incorrect amount of PVC found with selector %s", selector)
+				} else {
```
Comment (girishramnani) [inline pkg/kclient/volumes_test.go]: is this else needed? it seems to just be returning? which would happen anyway if the `if statement` is not executed.

**Comment by girishramnani** [comment]
a small nitpick @maysunfaisal :point_up:

File: pkg/kclient/volumes_test.go (hunk 5 above)
Comment (maysunfaisal) [reply pkg/kclient/volumes_test.go]: @girishramnani I've updated the commit, thx!

**Pull request merged**